yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts

```diff
@@ -0,0 +1,210 @@
+            `Cannot store index ${index} for secret ${secretStr} and txHash ${txHash.toString()}: ` +
+              `a different index ${existingEntry.index} already exists for this secret-txHash pair`,
+          );
+        }
+        // If it exists with the same index, ignore the update (no-op)
+      } else {
+        // If it doesn't exist, add it
+        await this.#pendingIndexes.set(secretStr, [...existing, { index, txHash: txHash.toString() }]);
+      }
+    }
+  }
+
+  /**
+   * Returns the transaction hashes of all pending transactions that contain indexes within a specified range
+   * for a given directional app tagging secret.
+   * @param secret - The directional app tagging secret to query pending indexes for.
+   * @param startIndex - The lower bound of the index range (inclusive).
+   * @param endIndex - The upper bound of the index range (exclusive).
+   * @returns An array of unique transaction hashes for pending transactions that contain indexes in the range
+   * [startIndex, endIndex). Returns an empty array if no pending indexes exist in the range.
+   */
+  async getTxHashesOfPendingIndexes(
+    secret: DirectionalAppTaggingSecret,
+    startIndex: number,
+    endIndex: number,
+  ): Promise<TxHash[]> {
+    const secretStr = secret.toString();
+    const existing = (await this.#pendingIndexes.getAsync(secretStr)) ?? [];
+    const txHashes = existing
+      .filter(entry => entry.index >= startIndex && entry.index < endIndex)
+      .map(entry => entry.txHash);
+    return Array.from(new Set(txHashes)).map(TxHash.fromString);
+  }
+
+  /**
+   * Returns the last (highest) finalized index for a given secret.
+   * @param secret - The secret to get the last finalized index for.
+   * @returns The last (highest) finalized index for the given secret.
+   */
+  getLastFinalizedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    return this.#lastFinalizedIndexes.getAsync(secret.toString());
+  }
+
+  /**
+   * Returns the last used index for a given directional app tagging secret, considering both finalized and pending
+   * indexes.
+   * @param secret - The directional app tagging secret to query the last used index for.
+   * @returns The last used index.
+   */
+  async getLastUsedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    const lastFinalizedIndex = await this.#lastFinalizedIndexes.getAsync(secret.toString());
+    const pendingTxScopedIndexes = (await this.#pendingIndexes.getAsync(secret.toString())) ?? [];
+    const pendingIndexes = pendingTxScopedIndexes.map(entry => entry.index);
+
+    if (pendingTxScopedIndexes.length === 0) {
+      return lastFinalizedIndex;
+    }
+
+    // As the last used index we return the highest one from the pending indexes. Note that this might not technically
+    // be the last used index in case we sent logs from multiple PXEs in parallel but that's just a detail.
+    return Math.max(...pendingIndexes);
```

Comment:
This assumes the pending indices will always be larger than the last finalized. This is wrong as transactions are not necessarily getting mined in the order in which we send them, e.g. index 1 is used in tx B with low fees, index 2 is used in tx B with high fees, B gets mined and eventually finalized, A sits on the mempool until tx fees drop.